src/boot/bootctl.c

```diff
@@ -1174,100 +1285,33 @@ static int install_loader_config(const char *esp_path) {
         return 1;
 }
 
-static int install_machine_id_directory(const char *root) {
+static int install_entry_directory(const char *root) {
         assert(root);
-        assert(arg_make_machine_id_directory >= 0);
+        assert(arg_make_entry_directory >= 0);
 
-        if (!arg_make_machine_id_directory)
+        if (!arg_make_entry_directory)
                 return 0;
 
-        assert(!sd_id128_is_null(arg_machine_id));
-        return mkdir_one(root, SD_ID128_TO_STRING(arg_machine_id));
+        assert(arg_entry_token);
+        return mkdir_one(root, arg_entry_token);
 }
 
-static int install_machine_info_config(void) {
-        _cleanup_free_ char *contents = NULL;
-        size_t length;
-        bool need_install_layout = true, need_machine_id;
+static int install_entry_token(void) {
         int r;
 
-        assert(arg_make_machine_id_directory >= 0);
-
-        /* We only want to save the machine-id if we created any directories using it. */
-        need_machine_id = arg_make_machine_id_directory;
-
-        _cleanup_fclose_ FILE *orig = fopen("/etc/machine-info", "re");
-        if (!orig && errno != ENOENT)
-                return log_error_errno(errno, "Failed to open /etc/machine-info: %m");
-
-        if (orig) {
-                _cleanup_free_ char *install_layout = NULL, *machine_id = NULL;
+        assert(arg_make_entry_directory >= 0);
+        assert(arg_entry_token);
 
-                r = parse_env_file(orig, "/etc/machine-info",
-                                   "KERNEL_INSTALL_LAYOUT", &install_layout,
-                                   "KERNEL_INSTALL_MACHINE_ID", &machine_id);
-                if (r < 0)
-                        return log_error_errno(r, "Failed to parse /etc/machine-info: %m");
-
-                rewind(orig);
-
-                if (!isempty(install_layout))
-                        need_install_layout = false;
+        /* Let's save the used entry token in /etc/kernel/entry-token if we used it to create the entry
+         * directory, or if anything else but the machine ID */
 
-                if (!isempty(machine_id))
-                        need_machine_id = false;
-
-                if (!need_install_layout && !need_machine_id) {
-                        log_debug("/etc/machine-info already has KERNEL_INSTALL_MACHINE_ID=%s and KERNEL_INSTALL_LAYOUT=%s.",
-                                  machine_id, install_layout);
-                        return 0;
-                }
-
-                r = read_full_stream(orig, &contents, &length);
-                if (r < 0)
-                        return log_error_errno(r, "Failed to read /etc/machine-info: %m");
-        }
+        if (!arg_make_entry_directory && arg_entry_token_type != ARG_ENTRY_TOKEN_MACHINE_ID)
+                return 0;
```

Comment:
So I figure we should change this to only write /etc/kernel/entry-token if an explicit entry-token was provided. We can find machine-id, os-id and image-os-id in kernel-install automatically without an additional file.

Reply:
Hmm, so the line you highlighted is actually borked. It should read:

```c
if (!arg_make_entry_directory && arg_entry_token_type == ARG_ENTRY_TOKEN_MACHINE_ID)
        return 0;
```

i.e the idea here was that we'll persist the it whenever we are told to create the dir or when it is set to any non-default value.

Why persist in even if the selected id is "os-id" or "image-os-id" instead of purely relying on auto-detection in kernel-install for those cases? Because unlike for the machine ID even if everything is set up OK there's a risk that two distinct images will use the same ID and fight for the directory (think: two images generated from Fedora might use the "fedora" entry token. we should never do that by accident, but only on request). Hence I figured we should not risk that: if a risky way to identify the entries is selected then we should remember that and cut any auto-detection short, and just do what we were told.

does that thinking make any sense?

(I mean, it would be Ok if there was exactly one other string it might be OK to rely on a logic of "machine-id" dir doesn't exist, hence use this one other string, but IRL there are at least three potential strings — os-id, os-image-id, "Default" – and possible more in future hence relying on searching here is icky I think)

Reply:
Shouldn't it be as follows then:

```
if (!arg_make_entry_directory || arg_entry_token_type == ARG_ENTRY_TOKEN_MACHINE_ID)
        return 0;
```
> Why persist in even if the selected id is "os-id" or "image-os-id" instead of purely relying on auto-detection in kernel-install for those cases? Because unlike for the machine ID even if everything is set up OK there's a risk that two distinct images will use the same ID and fight for the directory (think: two images generated from Fedora might use the "fedora" entry token. we should never do that by accident, but only on request). Hence I figured we should not risk that: if a risky way to identify the entries is selected then we should remember that and cut any auto-detection short, and just do what we were told.

Should we do searching in kernel-install at all then if we're always going to write /etc/kernel/entry-token? Why not remove the ickyness altogether and require /etc/kernel/entry-token to be written when using anything that's not the machine ID? If we're going to write to entry-token from bootctl anyway, we might as well require it and remove the heuristics from kernel-install.

Reply:
> Shouldn't it be as follows then:
>
> if (!arg_make_entry_directory || arg_entry_token_type == ARG_ENTRY_TOKEN_MACHINE_ID)
>        return 0;

No. Note that the condition is chekcing when to *not* create the dir. i.e. it's for when to exit early from the function. Or in other words, if you invert it, then you'll see under which conditions the dir is created:

```
if (arg_make_entry_directory || arg_entry_token_type != ARG_ENTRY_TOKEN_MACHINE_ID) {
…
}
```

i.e. we persist the token when we are told to create dir, or when the token is not the default of machine-id.

Reply:
> Should we do searching in kernel-install at all then if we're always going to write /etc/kernel/entry-token? Why not remove the ickyness altogether and require /etc/kernel/entry-token to be written when using anything that's not the machine ID? If we're going to write to entry-token from bootctl anyway, we might as well require it and remove the heuristics from kernel-install.

The main reason is compat: we used to search for "default", and we should probably continue to do so. I see "default" and ID=/IMAGE_ID= in the same boat there...

dunno, we could remove the searching altogether, i#d not be stronlgy opposed to that, but i think it does make some sense...